From 1f5696925ed2464c3a00af91eb009021645eb4af Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Wed, 20 Dec 2017 18:21:08 +0900
Subject: [PATCH 389/909] media: i2c: adv7604: Add suspend and resume support

commit e26341b641b46e101479f641f78973f15ca17664 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

In order to support suspend to ram, this patch registers the
resume and suspend function in adv7604 driver.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/i2c/adv7604.c | 42 +++++++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/drivers/media/i2c/adv7604.c b/drivers/media/i2c/adv7604.c
index 0e2ac20a9517..c3ee85e2cbbd 100644
--- a/drivers/media/i2c/adv7604.c
+++ b/drivers/media/i2c/adv7604.c
@@ -3658,9 +3658,51 @@ static int adv76xx_remove(struct i2c_client *client)
 
 /* ----------------------------------------------------------------------- */
 
+#ifdef CONFIG_PM_SLEEP
+static int adv76xx_suspend(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	u8 value;
+
+	value = io_read(sd, 0x0c) | (u8)0x20;
+	io_write(sd, 0x0c, value);
+
+	return 0;
+}
+
+static int adv76xx_resume(struct device *dev)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct v4l2_subdev *sd = i2c_get_clientdata(client);
+	struct adv76xx_state *state = to_state(sd);
+	int i;
+	int ret;
+
+	for (i = 1; i < ADV76XX_PAGE_MAX; ++i) {
+		if (!(BIT(i) & state->info->page_mask))
+			continue;
+		ret = io_write(sd, 0xf2 + i,
+			       state->pdata.i2c_addresses[i] << 1);
+		if (ret)
+			return ret;
+	}
+
+	return adv76xx_core_init(sd);
+}
+
+static SIMPLE_DEV_PM_OPS(adv76xx_pm_ops, adv76xx_suspend, adv76xx_resume);
+#define ADV76XX_PM_OPS (&adv76xx_pm_ops)
+#else
+#define ADV76XX_PM_OPS (NULL)
+#endif
+
+/* ----------------------------------------------------------------------- */
+
 static struct i2c_driver adv76xx_driver = {
 	.driver = {
 		.name = "adv7604",
+		.pm = ADV76XX_PM_OPS,
 		.of_match_table = of_match_ptr(adv76xx_of_id),
 	},
 	.probe = adv76xx_probe,
-- 
2.17.1

